#! /bin/sh /usr/share/dpatch/dpatch-run
## 20_fix_read_pdo_count.dpatch by  <cncmill@cncmill>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: fix read-pdo error with stoeber servo amp

@DPATCH@

diff -urNp ethercat-1.5.1.orig/master/fsm_pdo_entry.c ethercat-1.5.1/master/fsm_pdo_entry.c
--- ethercat-1.5.1.orig/master/fsm_pdo_entry.c	2012-01-10 13:03:28.000000000 +0100
+++ ethercat-1.5.1/master/fsm_pdo_entry.c	2012-03-15 09:06:57.615486443 +0100
@@ -216,7 +216,13 @@ void ec_fsm_pdo_entry_read_state_count(
         return;
     }
 
-    if (fsm->request.data_size != sizeof(uint8_t)) {
+    if (fsm->request.data_size == sizeof(uint8_t)) {
+      fsm->entry_count = EC_READ_U8(fsm->request.data);
+    } else if (fsm->request.data_size == sizeof(uint16_t)) {
+      fsm->entry_count = EC_READ_U16(fsm->request.data);
+    } else if (fsm->request.data_size == sizeof(uint32_t)) {
+      fsm->entry_count = EC_READ_U32(fsm->request.data);
+    } else {
         EC_SLAVE_ERR(fsm->slave, "Invalid data size %zu at uploading"
                 " SDO 0x%04X:%02X.\n",
                 fsm->request.data_size, fsm->request.index,
@@ -225,8 +231,6 @@ void ec_fsm_pdo_entry_read_state_count(
         return;
     }
 
-    fsm->entry_count = EC_READ_U8(fsm->request.data);
-
     EC_SLAVE_DBG(fsm->slave, 1, "%u PDO entries mapped.\n", fsm->entry_count);
 
     // read first PDO entry
